### ISSUE TITLE
Analyzer throws compile error whereas dart passes when two depending classes (B<X extends A<X>> and A<X extends B<X>>) are declared.

### ISSUE THREAD

**Issue by iarkh** [dart-lang/sdk#34628]
Dart SDK Version: 2.1.0-dev.6.0
OS: Windows 10 (64 bit)

The following sample code declares two parametrized classes which depend on each other:
``` dart
class B<X extends A<X>> {}
class A<X extends B<X>> {}

Type typeOf<X>() => X;

main() {
  print(typeOf<A>());
  print(typeOf<B>());
}
```
Dart passes here and prints A and B actual types:
> A<B<dynamic>>
> B<A<dynamic>>

Dartanalyzer throws a compile error:
> Analyzing test1.dart...
>   error - 'X' doesn't extend 'B<X>' at test1.dart:1:21 - type_argument_not_matching_bounds
>   error - 'X' doesn't extend 'A<X>' at test1.dart:2:21 - type_argument_not_matching_bounds
> 2 errors found.

Seems like dart behavior is correct.
And anyway, analyzer and runtime should behave in a similar way.

**Comment by stefantsov** [comment]
I think that in this case the behavior of Analyzer is correct.  CFE still misses some type checks.  It's currently being fixed (see https://github.com/dart-lang/sdk/issues/33308 and [CL 57621](https://dart-review.googlesource.com/c/sdk/+/57621)) and should land soon.

As for why it's a compile-time error, I believe, it's because `X` in the declaration of `B` doesn't extend `B<X>`, but is passed as a type argument to `A`.  Therefore, it should be rejected.  Similar reasoning applies to another `X` in the declaration of `A`.

**Comment by stefantsov** [comment]
Fixed as of aa6fb2d3a4298e60b5635f26507d5b39f4e531cb.

**Comment by eernstg** [comment]
Considering this snippet:

```dart
class B<X extends A<X>> {}
class A<X extends B<X>> {}
```

It is actually a **delicate question whether** that should be a compile-time **error** or an acceptable set of declarations, so I think it calls for a few extra thoughts.

We can expand a little on the available information and get something that is accepted by our tools:

```dart
class A<Y extends B<X, Y>, X extends A<Y, X>> {}
class B<X extends A<Y, X>, Y extends B<X, Y>> {}
class AA implements A<BB, AA> {}
class BB implements B<AA, BB> {}
class AB implements A<AB, AB>, B<AB, AB> {}

main() {
  B<Null, Null> ok1;
  A<Null, Null> ok2;
  A<BB, AA> ok3;
  B<AA, BB> ok4;
  A<AB, AB> ok5;
  B<AB, AB> ok6;
}
```

With a little extension we could get the same information crammed into a single type parameter (using the `&` that Java is using to avoid the same parsing problem):

```dart
class B<X extends A<X> & B<X>> {}
class A<X extends B<X> & A<X>> {}
```

Of course, you could make this nominal rather than structural (so clients will have to use a new type `AB` rather than just satisfying two separate subtype requirements), and then we're back in current Dart:

```dart
class A<X extends AB<X>> {}
class B<X extends AB<X>> {}
class AB<X extends AB<X>> implements A<X>, B<X> {}
```

But you could also claim that we can just use the available information in any case, reconsidering the original form:

```dart
class B<X extends A<X>> {}
class A<X extends B<X>> {}
```

During analysis, we'd need to assume that a client somewhere has already created an actual type argument `T` which will be the value for `X` in some context, and we are only going to accept `T` as an actual type argument because it satisfies the constraint `T <: A<T>`.

But that wouldn't be true unless `T` is an acceptable type argument for `A` in the first place, i.e., we must then also have `T <: B<T>`.

So from `X extends B<X>` (and the declaration of `B`) alone, we already know that every admissible value of `X` would satisfy both of the constraints that we need in order to accept the program.

In general, we could use such F-bounded declarations as sources of information about _what must hold_ for the given arguments, rather than just using the explicitly and locally declared relations (that is, when considering `A`: `X extends B<X>`) when we check whether each type is well-formed.

This generalization might be useful in practice, so **why not?** ;-)

The reason why **I think we should not do this** is that this process where we harvest information about 'what must hold' when we have an actual type argument might not terminate on its own, and that doesn't even require more than one class:

```dart
class A<X extends B<Y>, Y extends B<X>> {}
class B<X> {}
```

So, for checks on the type argument list of `C`, we know that any given actual type arguments `T` and `U` must satisfy `T <: B<U>` and `U <: B<T>`, so `T <: B<B<T>> <: B<B<B<B<T>>>> <: ...`.

So we wouldn't want to start some kind of inference engine to gather all logical consequences of a given set of class declarations. But we might still want to allow multiple types in a bound, which would presumably address pretty much the same design space, without the complexity explosion.
